### DISCUSSION TITLE
How do pointer arithmetic in viper

### DISCUSSION THREAD

**Discussion by GuillaumeLeclerc** [micropython/micropython#11640] [Using MicroPython]
Hello,

I'm trying to write a tight control loop in micropython. The code uses PIOs, inline assembly and viper and is plenty fast. The only issue is that garbage collection completely ruins the real time aspect of the application.

I need to store state and was using struct until I realized (#11639) that just accessing it in `viper` allocates memory for some reason.

My second option, (much less readable/maintanable) was to hard code offsets in my code and use `ptr32`. However it seems that viper will not let me do pointer arithmetic (addition). It complains that it can't add `ptr32` and `int`. What is the proper syntax to do that ?

**Comment by robert-hh** [comment]
> However it seems that viper will not let me do pointer arithmetic (addition). It complains that it can't add ptr32 and int
Would it help to do pointer indexing, `var_ptr[offset]` instead of `var_ptr + offset`?

**Reply by GuillaumeLeclerc** [reply]
It becomes very messy to not be able to pass arrays to functions, for example if I have nested structures I might have to pass multiple offsets and then inner functions have to do:

`var_ptr[OFFSET_1 + a * OFFSET_2]

That would make the code almost unreadable and hard to maintain.

**Reply by GitHubsSilverBullet** [reply]
My first thought:
A good C-Programmer can write C-Code in almost any programming language. ;-)

Then:
If you have to pass multiple offsets, then even C would have to use them in this particular way. Your `OFFSET_1 + a * OFFSET_2`  is in reality just an integer used as an index for `var_ptr[index] ` 
Passing slices of arrays will always allocate a new memory region. If you just read the data, a significantly faster memoryview might be helpful.

**Reply by rkompass** [reply]
You may avoid the ptr32 and int add by casting (like in C from pointer to uint before addition and from uint to pointer afterwards.
Not that the indexes correspond to multiples (1x, 2x or 4x) of bytes offset, depending on the size of the data the pointer addresse, ie. on ptr8, ptr16 or ptr32.

Example code:
```
@micropython.viper
def pointer_add1(ptr: ptr32, ofs1: uint, ofs2: uint) -> int:
    return ptr[ofs1+ofs2]

@micropython.viper
def pointer_add2(ptr: ptr32, ofs1: uint, ofs2: uint) -> int:
    pi1 = uint(ptr)+4*ofs1
    print('pi1:', pi1)
    p1 = ptr32(pi1)
    return p1[ofs2]

ap = addressof(a)
print('ap:', ap)

print(pointer_add1(ap, 2, 3))
print(pointer_add2(ap, 2, 3))
```
